src/control/character_controller.rs

```diff
@@ -546,28 +554,29 @@ impl KinematicCharacterController {
 
         let decomp = self.decompose_hit(translation_remaining, &hit.toi);
 
-        // An object is trying to slip if the tangential movement induced by its vertical movement
-        // points downward.
-        let slipping_intent = self.up.dot(&horiz_input_decomp.vertical_tangent) < 0.0;
+        // An object is trying to slip if its vertical (according to self.up) input motion points downward.
+        let slipping_intent = self.up.dot(&_vertical_input) < 0.0;
+
+        // An object is slipping if the tangential movement induced by its vertical movement points downward.
         let slipping = self.up.dot(&decomp.vertical_tangent) < 0.0;
 
-        // An object is trying to climb if its indirect vertical motion points upward.
-        let climbing_intent = self.up.dot(&input_decomp.vertical_tangent) > 0.0;
-        let climbing = self.up.dot(&decomp.vertical_tangent) > 0.0;
+        // An object is trying to climb if its vertical (according to self.up) input motion points upward.
+        let climbing_intent = self.up.dot(&_vertical_input) > 0.0;
```

Comment:
part 2 of the fix: using input to infer "user intent", rather than movement tangent.

This leads to unused `horiz_input_decomp` and `input_decomp`.

Reply:
Mmh, I think I tried that at some point and this had a side-effect of having the user climb some slopes it shouldn’t. It would be worth moving around in the character-controller scene.

This also makes me think that perhaps we should make a test scene where we have many character-controller (in isolation) that try to walk on various surface configurations. That would help spot regressions more easily without having to move the character manually.

Reply:
a scene with multiple character controller is an interesting idea ; to be noted I added an automated test for character controller.